aioxmpp/pep/service.py

```diff
@@ -0,0 +1,363 @@
+    callback registered! It is strongly recommended to explicitly
+    close the registered node if it is no longer needed or to use the
+    :class:`register_pep_node` descriptor for automatic life-cycle
+    handling.
+
+    .. signal:: on_item_publish(jid, node, item, message=None)
+
+       Fires when an event is received for this PEP node. The arguments
+       are as for :attr:`aioxmpp.PubSubClient.on_item_publish`.
+
+    .. autoattribute:: notify
+
+    .. autoattribute:: registered_feature
+
+    .. automethod:: close
+
+    """
+
+    def __init__(self, pep_service, node, register_feature, notify):
+        self._pep_service = pep_service
+        self._node = node
+        self._feature_registered = register_feature
+        self._notify = notify
+        self._closed = False
+
+        if self._feature_registered:
+            self._register_feature()
+
+        if self._notify:
+            self._register_notify()
+
+    on_item_publish = callbacks.Signal()
+
+    def _register_feature(self):
+        self._pep_service._disco_server.register_feature(self._node)
+        self._feature_registered = True
+
+    def _unregister_feature(self):
+        self._pep_service._disco_server.unregister_feature(self._node)
+        self._feature_registered = False
+
+    def _register_notify(self):
+        self._pep_service._disco_server.register_feature(self._notify_feature)
+        self._notify = True
+
+    def _unregister_notify(self):
+        self._pep_service._disco_server.unregister_feature(
+            self._notify_feature)
+        self._notify = False
+
+    def _unregister(self):
+        if self._notify:
+            self._unregister_notify()
+
+        if self._feature_registered:
+            self._unregister_feature()
+
+    def close(self):
+        """
+        Unclaim the PEP node and unregister the registered features.
+
+        It is not necessary to call close if this claim is managed by
+        :class:`register_pep_node`.
+        """
+        if self._closed:
+            return
+
+        self._closed = True
+        self._pep_service._unclaim(self.node_namespace)
+        self._unregister()
+
+    @property
+    def node_namespace(self):
+        """The claimed node namespace"""
+        return self._node
+
+    @property
+    def _notify_feature(self):
+        return self._node + "+notify"
+
+    @property
+    def notify(self):
+        """
+        Whether we have enabled the ``+notify`` feature to automatically
+        receive notifications.
+
+        When setting this property the feature is registered and
+        unregistered appropriately.
+        """
+        return self._notify
+
+    @notify.setter
+    def notify(self, value):
+        if self._closed:
+            raise RuntimeError(
+                "modifying a closed RegisteredPEPNode is forbidden"
+            )
+        # XXX: do we want to do strict type checking here?
+        if (not value) == self._notify:
```

Comment:
My point here is actually about two things: first, that ``not value`` and second that it’s clearer to if-and-return than if-and-do-things if possible.

I have no strong opinion on bool-ising self._notify.